fuel-derive/src/serialize.rs

```diff
@@ -0,0 +1,495 @@
+                #a * #b
+            }),
+        }
+    }
+}
+
+/// Determines serialized size of a type using `mem::size_of` if possible.
+fn try_builtin_sized(ty: &syn::Type) -> Option<TypeSize> {
+    match ty {
+        syn::Type::Group(group) => try_builtin_sized(&group.elem),
+        syn::Type::Array(arr) => {
+            let elem_size = try_builtin_sized(arr.elem.as_ref())?;
+            let elem_count = TypeSize::from_expr(&arr.len);
+            Some(elem_size * elem_count)
+        }
+        syn::Type::Tuple(tup) => tup
+            .elems
+            .iter()
+            .map(try_builtin_sized)
+            .fold(Some(TypeSize::Constant(0)), |acc, item| Some(acc? + item?)),
+        syn::Type::Path(p) => {
+            if p.qself.is_some() {
+                return None
+            }
+
+            if !is_sized_primitive(p.path.get_ident()?.to_string().as_str()) {
+                return None
+            }
+
+            Some(TypeSize::Computed(quote! {
+                fuel_types::canonical::aligned_size(::core::mem::size_of::<#p>())
+            }))
+        }
+        _ => {
+            panic!("Unsized type {:?}", ty);
+        }
+    }
+}
+
+fn constsize_fields(fields: &syn::Fields) -> (TokenStream2, TokenStream2) {
+    let mut sizes: Vec<_> = fields
+        .iter()
+        .filter_map(|field| {
+            let type_ = &field.ty;
+            if should_skip_field(&field.attrs) {
+                None
+            } else if let Some(size_code) = try_builtin_sized(type_) {
+                Some(size_code)
+            } else {
+                Some(TypeSize::Computed(quote! {
+                    <#type_>::SIZE_STATIC
+                }))
+            }
+        })
+        .collect();
+    sizes.sort_by(|a, b| a.partial_cmp(b).unwrap());
+    let static_size: TokenStream2 = sizes.into_iter().fold(quote! { 0 }, |acc, item| {
+        quote! {
+            #acc + #item
+        }
+    });
+
+    let no_dynamic_size: TokenStream2 = fields
+        .iter()
+        .filter_map(|field| {
+            let type_ = &field.ty;
+            if should_skip_field(&field.attrs) || try_builtin_sized(type_).is_some() {
+                return None
+            }
+            Some(quote! {
+                <#type_>::SIZE_NO_DYNAMIC
+            })
+        })
+        .fold(quote! { true }, |acc, item| {
+            quote! {
+                #acc && #item
+            }
+        });
+
+    (static_size, no_dynamic_size)
+}
+
+/// Derives `Serialize` trait for the given `struct` or `enum`.
+pub fn serialize_derive(mut s: synstructure::Structure) -> TokenStream2 {
+    s.add_bounds(synstructure::AddBounds::Fields)
+        .underscore_const(true);
+
+    let serialize = match s.ast().data {
+        syn::Data::Struct(_) => serialize_struct(&s),
+        syn::Data::Enum(_) => serialize_enum(&s),
+        _ => panic!("Can't derive `Serialize` for `union`s"),
+    };
+
+    // crate::utils::write_and_fmt(
```

Comment:
Should this be removed?

Reply:
Maybe? That's the main way to debug derive macros, so maybe I'll rather just clean it up a bit and document what it does instead.